### ISSUE TITLE
IVP loss with 3d images

### ISSUE THREAD

**Issue by benjijamorris** [kevinjohncutler/omnipose#44]
As of now, torchvf only works in 2d, so I think the loss calculation in its current form doesn't work for 3d inputs

**Comment by kevinjohncutler** [comment]
Oops, that is an oversight on my part. I am still investigating how `ivp_loss()` improves training, but the enhancements are subtle. You can definitely revert the `core.loss()` function to its previous state in the meantime. I'll invite @ryanirl to chime in too, but it looks like it should be pretty straightforward to generalize to ND (unless we run into some pytorch functions that limit us to 3D). 

In ivp_loss.py from the losses directory, we have
1. torch.meshgrid: this works for ND. 
2. Any instances like `y, x = torch.meshgrid` or `B, C, H, W = vf.shape` need to be changed to `g = torch.meshgrid` or `s = vf.shape`. Then `g` can be stacked in the required order via `g[-1::]` and B can just be indexed as s[0] etc. 

ivp_loss requires interp_vf and ivp_solver from the numerics directory, and I see that the real work will be in functional.py under the interpolation subdirectory. All those calculations are similarly parsed out in terms of x and y, but `torch.lerp()` will not pose an issue. The indexing will just take a little bit to figure out. Fortunately, looks like the solvers are 'vectorized' and do not require any modification.

@benjijamorris let me know if you already dug into the details or if I should go ahead and make the required alterations.

**Comment by ryanirl** [comment]
I would need some sample 3D data to experiment with this. But right away, the bilinear interpolation will need some work most likely if you want to generalize it to 3D. I had to do do some _not so easily readable_ PyTorch tricks to really vectorize the code and crunch out performance so training can be quick. I have limited bandwidth right now, but would be able to take a look at it in a week or so once summer starts. The IVP solvers _might_ generalize to 3D, although if they don't, those should be the easiest to get up and running.

**Comment by benjijamorris** [comment]
Awesome thanks! Just wanted to bring this to your attention. Can confirm that torchvf generalizes nicely to 3d cases, still sorting out how to handle anisotropy for my use case. Thanks @kevinjohncutler, @ryanirl for your awesome repos!

**Comment by kevinjohncutler** [comment]
@ryanirl what is the reasoning behind not using `torch.nn.functional.grid_sample()` for interpolation? Is it for performance, or perhaps [reproducibility](https://pytorch.org/docs/stable/generated/torch.nn.functional.grid_sample.html)? It is also only fit for 2D and 3D (4D and 5D tensors) so I am inclined to just adapt your code to ND.

**Comment by kevinjohncutler** [comment]
@ryanirl One more reason not to use `torch.nn.functional.grid_sample()` is that it is not supported on the MPS backend. 

This brought me back to another change I made to post-processing a while ago, and that's removing bilinear interpolation in favor of nearest neighbor interpolation for mask reconstruction. The reason is that opposing components in the field average to zero, and so the effective flow field magnitude near cell-cell boundaries - and thus pixel movement at the most critical parts of the image - can sometimes be reduced when interpolated. It's pretty rare, and mostly only affects boundary pixels whose flows are tangent to the boundary (the network tends to interpolate < or > as ^ or v rather than 0). Ivp_loss seems to help avoid those (literal) edge cases, just as I hoped. 

I think the same reasoning applies to your loss function, and the benefit is that it is far easier to implement. I see you have it set to do bilinear_interpolation_batched by default, so I made nearest_interpolation_batched to replace it in ND. Seems to work well, but I am curious if you know or anticipate any issues that could arise from using nearest neighbor instead of bilinear interpolation in this context.

**Comment by ryanirl** [comment]
I agree that speed and ease of implementation are convincing upsides. Although the downside to nearest neighbor interpolation would be reduced integration performance before the clustering step. I wish I had an example on hand, but I often found bilinear interpolation to work better in a ground truth sense over nearest neighbor interpolation, which I found to sometimes have _choppy_ results after interpolation. To verify or disprove this, maybe consider running a ground truth validation with both interpolation schemes and see if the mAP IoU changes.

**Comment by kevinjohncutler** [comment]
So my results so far (training and testing on BPCIS) indicate that ivp_loss is slowing down training a lot, as even 1.5x the number of epochs is not getting me to the same level of performance (the time per epoch is not really a concern at the moment). However, I had a step size of 0.2 and the number of steps set at 15. Because the flow field in Omnipose is multiplied by 5 for prediction balancing, this effectively means that the trajectories will be 15px long. The first problem with this is that it is too long for many cells in the dataset (would capture a lot of bounding around in the middle of the cell that is irrelevant). The second is that it is redundant - in Omnipose at least, pixels will follow each other along the unit magnitude flow field to the cell skeleton, so there is no need to trace more than one pixel of movement (maybe two to be sure there is some overlap). The model I am training now is tracing 2px of movement, and it already looks to be converging faster than the 15-step model. I should know in a couple days if we are able to recapitulate the original Omnipose results (or improve them) using these new settings for ivp_loss.

**Comment by ryanirl** [comment]
Sounds good, let me know what you happens in your experiments. In tuning the IVP loss, I tried to personally use similar total trajectory length (`dx` * `n_steps`) in the training as in the evaluation (found through GT validation). Although during training I found it helpful to reduce `n_steps` and increase `dx` to speed it up. In my experiments I used `dx = 0.25` and `n_steps = 8` during training and `dx = 0.1` and `n_steps = 25` during testing of the bact_phase subset. So you can see the tradeoff in `n_steps`. Although given that Omnipose flow fields aren't smoothed or reduced near the center (last time I checked), I would assume that setting `dx` to be large would be redundant as you mentioned.

**Comment by kevinjohncutler** [comment]
Sorry for the delay, I got side tracked with lab work and then _obsessed_ with optimizing the training loop. As a rough metric, a training run that used to take 2.5 days now trains in <5hr.  

On a bit of a tangent: I have long wondered why torchvf was reported as much faster at training, and I think the answer was the augmentation bottleneck rather than the ivp_loss function itself. Considering the results from Ryan's paper, one could argue that my augmentations are unnecessarily complicated and thus could be simplified in favor of speed. However, that might come at the cost of generality, which is hard to quantify without a test dataset that is a lot more diverse than the training set (not at all the case with BPCIS). 

My tests so far show that my nearest_batched version gives as good or better performance compared to the default bilinear_batched. I still need to test on 3D, but once I confirm that works, I will submit a PR.

**Comment by kevinjohncutler** [comment]
@benjijamorris you can try out the most recent github version for 3D training using torchvf (and you will enjoy the massive speedups). Be sure to use my fork of torchvf until the main branch is updated. 

@ryanirl, let me know when you have time to review that pull request and push it to PyPi. Once that is done, I can make a new Omnipose release using it. No rush though, just want to plan ahead.

**Comment by ryanirl** [comment]
I'll plan to look it over it detail on Sunday.
